### PR TITLE
snap: delete stale snapshot immediately

### PR DESCRIPTION
**Pull request by BusyJay**
When replicating a peer, If it splits frequently, its snapshot may be always overlapped with other regions, so the snapshots will never be applied. Unfortunately, GC check can do nothing about the it, because the snapshot is never applied so `compacted_index` is always 0 which is less than any valid snapshot's index. And it's hard to check it in GC too. Because you can't tell if the snapshot is verified to be stale or just still buffered in the channel.

So this pr deletes the snapshot files immediately once it fails the snapshot check.

### CODE REVIEW

**Comment by siddontang** [comment]
CI failed

**Comment by siddontang** [comment]
any update @BusyJay

**Comment by siddontang** [comment]
can we use failpoint for this test now?

**Comment by BusyJay** [comment]
I will add a test case once #2474 is merged, which introduces failpoint cases into integration tests.

**Comment by BusyJay** [comment]
PTAL

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @overvenus

File: src/raftstore/store/worker/region.rs [hunk 1]
```diff
@@ -109,6 +109,7 @@ impl SnapContext {
             &raw_snap,
             region_id
         ));
+        fail_point!("handle_gen", region_id == 1, |_| Ok(()));
```
Comment (overvenus) [inline src/raftstore/store/worker/region.rs]: Better adding some comments, I have trouble understanding it.

**Pull request merged**